Adds the pipelines:diff command

krummi opened the pull request:
This PR implements the "heroku pipelines:diff" command.

Notes on implementation:
* Got info on the undocumented kolkrabbi endpoints from a Herokai (thanks @gudmundur).
* Tried to stick to conventions in this project (i.e. regarding what to use from ES6 etc).
* Added two dependencies; lodash and request. They're probably easy to get rid of, if you guys want that.
* There are currently no tests for this, we should probably add some.

First time doing ES6, so review those parts carefully. ;)

appleton commented on commands/pipelines/diff.js:
```diff
@@ -0,0 +1,157 @@
+  }).get(1);
+
+  console.log(`\n${sourceApp.name} is ahead of ${downstreamApp.name} by ${githubDiff.ahead_by} commit${githubDiff.ahead_by === 1 ? '' : 's'}:`);
+  for (let i = githubDiff.commits.length - 1; i >= 0; i--) {
```
By chance I tested this on a pipeline where the commit sha of the production app is no longer available on GitHub - I think I must've rebased and force pushed or something. Anyway, that case causes an exception on this line:

```sh
TypeError: Cannot read property 'length' of undefined
    at diff (/Users/aappleton/projects/heroku-pipelines/commands/pipelines/diff.js:67:34)
    at next (native)
    at onFulfilled (/Users/aappleton/projects/heroku-pipelines/node_modules/heroku-cli-util/node_modules/co/index.js:65:19)
    at tryCatcher (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:507:31)
    at Promise._settlePromiseAt (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:581:18)
    at Promise._settlePromises (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:697:14)
    at Async._drainQueue (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:368:17)
```

krummi replied:
Ah, good catch!

appleton commented:
Thanks for the contribution! I'm making a list of cases which I'll run through to test:

- [x] App and all downstream apps are connected to GitHub
- [ ] App is connected to GitHub, some downstream apps are not
- [ ] App is connected to GitHub, all downstream apps are not
- [ ] App is not connected to GitHub

appleton commented on commands/pipelines/diff.js:
```diff
@@ -0,0 +1,157 @@
+    // hashes differ from the target hash.
+    const uniqueDownstreamHashes = _.uniq(_.pluck(downstreamHashes, 'hash'));
+    if (uniqueDownstreamHashes.length === 1 && uniqueDownstreamHashes[0] === targetHash.hash) {
+      console.log(`\nEverything is up to date.`);
```
I think it would be good to print out which apps were compared here. This will give the user a little more feedback in the case that some of the apps they tried to compare are not connected to GitHub. How about:

```
my-app-staging is up to date with my-app-production
my-app-staging was not compared to my-app-production as my-app-production is not connected to GitHub
```

If this is tricky to implement for some reason then a simpler message is also fine but I think we should make the GitHub connection limitation clear.

appleton commented on .gitignore:
```diff
@@ -0,0 +1 @@
+node_modules/
```
:) thanks

appleton commented on commands/pipelines/diff.js:
```diff
@@ -0,0 +1,157 @@
+
+function* diff(sourceApp, downstreamApp, repo, githubToken, herokuUserAgent) {
+  if (sourceApp.hash === downstreamApp.hash) {
+    console.log(`\neverything is up to date between ${sourceApp.name} and ${downstreamApp.name}`);
```
Could we use `cli.log` here please? It integrates better with the test suite.

appleton commented on commands/pipelines/diff.js:
```diff
@@ -0,0 +1,157 @@
+      headers: { 'Accept': PIPELINES_HEADER }
+    });
+  } catch (err) {
+    console.log(err);
```
You could change this to a `cli.debug` statement. That way it'll be hidden for most users.

krummi commented:
Awesome - very thorough review. Will try to address the things you mention soon.

krummi commented:
I've updated the PR to reflect your comments @appleton. It now looks much more carefully at the GitHub connections of the target/downstream apps as you mentioned and provides error reporting on more edge cases than before. I've also added a pretty thorough test-suite for the command.

appleton commented:
This is great! Thanks so much for this and extra special thanks for adding the tests :100:

appleton commented:
Published to npm as v0.27.0. Thanks again! :beers:

krummi commented:
Fantastic and glad to help! :+1:
